pkg/kube/krt/core.go

```diff
@@ -0,0 +1,196 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"fmt"
+	"reflect"
+
+	"google.golang.org/protobuf/proto"
+
+	"istio.io/istio/pkg/kube/controllers"
+	istiolog "istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/ptr"
+	"istio.io/istio/pkg/slices"
+)
+
+var log = istiolog.RegisterScope("krt", "")
+
+type Collection[T any] interface {
+	GetKey(k Key[T]) *T
+	List(namespace string) []T
+	Register(f func(o Event[T]))
+	RegisterBatch(f func(o []Event[T]))
+}
+
+type Singleton[T any] interface {
+	Get() *T
+	Register(f func(o Event[T]))
+	AsCollection() Collection[T]
+}
+
+func batchedRegister[T any](c Collection[T], f func(o Event[T])) {
+	c.RegisterBatch(func(events []Event[T]) {
+		for _, o := range events {
+			f(o)
+		}
+	})
+}
+
+type erasedCollection interface {
+	register(f func(o []Event[any]))
+	hash() string
+}
+
+type erasedCollectionImpl struct {
+	r func(f func(o []Event[any]))
+	h string
+}
+
+func (e erasedCollectionImpl) hash() string {
+	return e.h
+}
+
+func (e erasedCollectionImpl) register(f func(o []Event[any])) {
+	e.r(f)
+}
+
+func eraseCollection[T any](c Collection[T]) erasedCollection {
+	return erasedCollectionImpl{
+		h: fmt.Sprintf("%p", c),
```

Comment:
Are we certain that the address of the collection is a suitable hash or is this something that maybe needs extra consideration?

Reply:
Nope :-) removed this